src/Components/Web.Extensions/src/InputFile/InputFile.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Web.Extensions
+{
+    /// <summary>
+    /// A component that wraps the HTML file input element and exposes a <see cref="Stream"/> for each file's contents.
+    /// </summary>
+    public class InputFile : ComponentBase, IInputFileJsCallbacks, IDisposable
+    {
+        private ElementReference _inputFileElement;
+
+        private IJSRuntime _jsRuntime = default!;
+
+        private IJSUnmarshalledRuntime? _jsUnmarshalledRuntime;
+
+        private InputFileJsCallbacksRelay? _jsCallbacksRelay;
+
+        [Inject]
+        private IServiceProvider ServiceProvider { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the event callback that will be invoked when the collection of selected files changes.
+        /// </summary>
+        [Parameter]
+        public EventCallback<IFileListEntry[]> OnChange { get; set; }
```

Comment:
As part of my programme of recommending APIs that we can extend in the future, can I suggest typing this as:

```suggestion
        public EventCallback<FileInputChangeEventArgs> OnChange { get; set; }
```

... with `FileInputChangeEventArgs` being some type (probably a `class` in this case as it's not a high-frequency event) that initially just exposes an `IReadOnlyList<IFileListEntry>`?

Theoretically there could be other context info we want to add to this in the future.